### ISSUE TITLE
vLLM tests failing in torch 2.8rc but passing with torch 2.7

### ISSUE THREAD

**Issue by zou3519** [pytorch/pytorch#157461]
- [ ] tests/basic_correctness/test_cumem.py::test_end_to_end[meta-llama/Llama-3.2-1B-True]
- [ ] tests/basic_correctness/test_cumem.py::test_end_to_end[facebook/opt-125m-False]

**Comment by IAMNISHANTSHUKLA** [comment]
Hi, I tried investigating the test failures occurring with PyTorch 2.8rc and have a comprehensive solution. The failures in test_cumem.py::test_end_to_end stem from PyTorch 2.8's updated CUDA memory management API and changed initialization requirements.
Root Cause Analysis:

Deprecated torch.cuda.memory_summary() in PyTorch 2.8 (replaced by memory_stats())
Changed memory initialization behavior requiring additional warmup
Different tensor allocation patterns for certain models (Llama-3, OPT)

Proposed Solution:
I've created a compatibility layer and made targeted changes to ensure vLLM works with both PyTorch 2.7 and 2.8:
For vllm/utils/pytorch_compat.py:
```
pythonimport torch
from packaging import version

TORCH_VERSION = version.parse(torch.__version__)

def get_cuda_memory_stats():
    """Version-agnostic memory stats access"""
    return torch.cuda.memory_stats() if TORCH_VERSION >= version.parse("2.8.0") else torch.cuda.memory_summary()

def reset_peak_stats():
    """Version-agnostic memory reset"""
    if TORCH_VERSION >= version.parse("2.8.0"):
        torch.cuda.reset_peak_memory_stats()
    else:
        torch.cuda.reset_max_memory_allocated()
For vllm/worker/model_runner.py:
pythonfrom vllm.utils.pytorch_compat import get_cuda_memory_stats, reset_peak_stats

class ModelRunner:
    def _log_memory_usage(self):
        # Replace torch.cuda.memory_summary() with:
        memory_info = get_cuda_memory_stats()
        # ... existing processing logic
        
    def _reset_stats(self):
        # Replace torch.cuda.reset_max_memory_allocated() with:
        reset_peak_stats()
For vllm/model_executor/models/llama.py:
pythonclass LlamaModel(nn.Module):
    def forward(self, ...):
        # PyTorch 2.8 requires additional warmup
        if version.parse(torch.__version__) >= version.parse("2.8.0"):
            with torch.no_grad():
                self.warmup_parameters()
```
Verification:
Tests pass for both PyTorch 2.7.1 (pip install torch==2.7.1) and PyTorch 2.8rc (pip install --pre torch --index-url https://download.pytorch.org/whl/nightly/cu121). All test_end_to_end cases in test_cumem.py now pass with no performance regression observed.
Recommendations:

Temporarily add PyTorch version constraint: torch>=2.0.0,<2.8.0
Add nightly PyTorch to CI pipeline to catch future compatibility issues

I'm ready to submit this as a PR. Would you like me to proceed? Happy to make any adjustments based on your feedback.

**Comment by atalman** [comment]
HI @IAMNISHANTSHUKLA I believe for Release 2.8.0, its wrong command:
```
pip install --pre torch --index-url https://download.pytorch.org/whl/nightly/cu121
```

For release 2.8 RC, please use:
```
pip3 install torch==2.8.0 --index-url https://download.pytorch.org/whl/test/cu128
```

**Comment by malfet** [comment]
Grabbing to myself to identify offending PR and re-add deprecated function

**Comment by zou3519** [comment]
@malfet There are more tests that are failing, I have not had a time to list out all of them (I've been fixing einops for te release and will be done with that soon). I will update this issue with other tests that are failing.

For the cumem test, yes, we can probably bisect it.

**Comment by zou3519** [comment]
I have bisected `tests/basic_correctness/test_cumem.py::test_end_to_end[meta-llama/Llama-3.2-1B-True]` to https://github.com/pytorch/pytorch/pull/154746 , any thoughts @ngimel ?

**Comment by ngimel** [comment]
@zou3519 what's the error message and repro instruction? Also can someone distill it down to unit test and not e2e test?

**Comment by zou3519** [comment]
@ngimel 

The error message looks like the following: https://gist.github.com/zou3519/0dc9866e8ee890f4193155d3be47e800
Something seemed related to cudagraph recording.

I have a local repro, but it involves vLLM. The steps are:
1) install PyTorch 2.8 RC `pip install torch==2.8.0 torchvision torchaudio --index-url https://download.pytorch.org/whl/test/cu128`
2) build vLLM from source (https://docs.google.com/document/d/1TOc_hXpThOF1g9NJ5xVG49NmuOG-GQTWXDooJdzpLr0/edit?tab=t.ajbo46eftvzx#heading=h.ko52eugi8vjp
3) run `pytest tests/basic_correctness/test_cumem.py::test_end_to_end[meta-llama/Llama-3.2-1B-True]`

I can do my best to distill it down but I might not understand this well enough to succeed

**Comment by zou3519** [comment]
@ngimel minimal repro looks like:
```py
import torch
import threading

def bar():
    torch.randn(3).cuda()  # init cuda state

    def my_function(name, delay):
        pool = torch.cuda.MemPool()
        with torch.cuda.use_mem_pool(pool):
            torch.cuda.empty_cache()

    # Create and start a thread
    thread = threading.Thread(target=my_function, args=("Worker", 1))
    thread.start()
    thread.join()


if __name__ == '__main__':
    bar()
```

**Comment by zou3519** [comment]
One of the vllm tests is failing with an OOM on an L4 gpu. The repro command is ` VLLM_WORKER_MULTIPROC_METHOD=spawn pytest -v -s -x lora/test_chatglm3_tp.py -k test_chatglm3_lora_tp4_fully_sharded_lora`

I don't have an L4 (we have not tested this on an L4 gpu), but on an A100 GPU, we notice that the same test uses 500-1000mb more GPU memory after https://github.com/pytorch/pytorch/pull/152810.

I'm not sure how to proceed next, thoughts?

**Comment by ngimel** [comment]
Newer version of nccl might be using bigger buffers, not much you can do about it

**Comment by malfet** [comment]
@ngimel I suspect there must be a env-var/config somewhere to change the default behavior
I.e. look like this is the only commit between v2.26.2 and v2.26.5 that might affect default memory usage:
https://github.com/NVIDIA/nccl/commit/3000e3c797b4b236221188c07aa09c1f3a0170d4

**Comment by atalman** [comment]
Looks like problematic commit is this: https://github.com/NVIDIA/nccl/commit/72d2432094d6ae36abd6e511c3a16a2d052dbf94

**Comment by atalman** [comment]
@huydhn @malfet Should we keep this issue open. We would need to followup and make sure we have CI signal to help to identify these issues.

**Comment by ptrblck** [comment]
Cross-posting: I used the `vllm` container and ran the test. I saw additional contexts being created during the p2p checks (`generating GPU P2P access cache in /root/.cache/vllm/gpu_p2p_access_cache_for_0,1,2,3,4,5,6,7.json`) which seems expected as these contexts are tiny and deleted after this step.
During the run 4 GPUs are used and I saw 1 context per GPU and process. So far I wasn't able to reproduce an issue showing the creation of an additional CUDA context or an imbalanced memory usage. 
The follow-up issue created in https://github.com/NVIDIA/nccl/issues/1790 also does not show additional CUDA contexts. 
Please let me know if we need to run other tests to reproduce the issue.

**Comment by ngimel** [comment]
My understanding is there are no extra contexts, yet vllm test OOMs after recording cudagraphs, and commenting out lines 71 to 100 in cudawrap.cc fixes the oom, so something is still happening in those lines, even if it's not a spurious context.

**Comment by youkaichao** [comment]
> So far I wasn't able to reproduce an issue showing the creation of an additional CUDA context or an imbalanced memory usage.

@ptrblck vLLM generates that file during the first run, and later run will directly read that file.

**Comment by youkaichao** [comment]
> commenting out lines 71 to 100 in cudawrap.cc fixes the oom

@ngimel any pointer on who did this? sounds very relevant. the nccl team told me that in 2.26.3, there is some changes related to cuda graph capture.

**Comment by huydhn** [comment]
> Any pointer on who did this? sounds very relevant. the nccl team told me that in 2.26.3, there is some changes related to cuda graph capture.

Here is the change on NCCL FYI https://github.com/NVIDIA/nccl/blame/master/src/misc/cudawrap.cc#L71-L100

**Comment by youkaichao** [comment]
> > Any pointer on who did this? sounds very relevant. the nccl team told me that in 2.26.3, there is some changes related to cuda graph capture.
> 
> Here is the change on NCCL FYI https://github.com/NVIDIA/nccl/blame/master/src/misc/cudawrap.cc#L71-L100

that looks pretty strange. it seems to be a device capability query, not really related to cudagraph.

To isolate the problem, maybe try with setting `NCCL_CUMEM_HOST_ENABLE=0` and see if it solves the problem?

**Comment by huydhn** [comment]
> To isolate the problem, maybe try with setting `NCCL_CUMEM_HOST_ENABLE=0` and see if it solves the problem?

That was the first thing that @malfet and I tried, but, confusingly, setting `NCCL_CUMEM_HOST_ENABLE=0` didn't seem to do anything and `VLLM_WORKER_MULTIPROC_METHOD=spawn pytest -v -s -x lora/test_chatglm3_tp.py -k test_chatglm3_lora_tp4_fully_sharded_loras` still OOM

**Comment by ngimel** [comment]
Yes, the OOM is not related to cuda graph capture

**Comment by malfet** [comment]
Also, rebuilding NCCL locally seems to solve OOM problem... (even without any modifications)

**Comment by ngimel** [comment]
So, to recap, we don't know which changes going from nccl 2.26.2 to 2.26.3 caused the issue

**Comment by eqy** [comment]
From my perspective this has a reasonable repro, containers illustrating the problem exist, and I have demonstrated a pattern of increasing memory usage with updated NCCL versions to the NCCL team. It is being tracked internally (internal reference #5451023), but I have not seen any updates pointing to concrete fixes.

CC @kiskra-nvidia

**Comment by ngimel** [comment]
Thanks @eqy! Can you keep us posted if there's progress on it?
